### PR TITLE
Fix IFTI with named arguments

### PR DESCRIPTION
**Pull request by dkorpel**
`deduceFunctionTemplateMatch` makes the assumption that the arguments and parameters arrays are parallel, and `resolveNamedArguments` doesn't handle template tuple parameters yet. Also, default parameters can come before `nfargs` now, so the logic comparing `argi` with `nfargs` has to be rewritten.

### CODE REVIEW

**Comment by dkorpel** [comment]
I think this is ready to go. There's future improvements for error messages and tuples, but I don't want this PR to grow bigger, and all the recent refactoring commits keep pulling out the rug from under this PR so I'd like to get this in before I have to rebase again :)

File: compiler/src/dmd/dtemplate.d [hunk 1]
```diff
@@ -3974,14 +3987,32 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
                 {
                     Parameter fparam = fparameters[j];
                     if (fparam.storageClass & STC.autoref)       // if "auto ref"
                     {
-                        Expression farg = fargs && j < fargs.length ? (*fargs)[j] : fparam.defaultArg;
+                        Expression farg = (j < args.length) ? args[j] : fparam.defaultArg;
+                        // resolveNamedArgs strips trailing nulls / default params
+                        // when it doesn't anymore, the ternary can be replaced with:
+                        // assert(j < resolvedArgs.length);
+                        if (!farg)
```
Comment (RazvanN7) [inline compiler/src/dmd/dtemplate.d:4014]: This `if` block can be merged with the next one.

File: compiler/src/dmd/dtemplate.d [hunk 2]
```diff
@@ -5723,8 +5754,7 @@ extern (C++) final class TemplateMixin : TemplateInstance
 
 /************************************
  * This struct is needed for TemplateInstance to be the key in an associative array.
- * Fixing https://issues.dlang.org/show_bug.cgi?id=15812 and
- * https://issues.dlang.org/show_bug.cgi?id=15813 would make it unnecessary.
+ * Fixing https://issues.dlang.org/show_bug.cgi?id=15813 would make it unnecessary.
```
Comment (RazvanN7) [inline compiler/src/dmd/dtemplate.d:5757]: This change, although correct, seems unrelated. I just want to make sure this change is intended and not the left overs of a different branch.

File: compiler/test/fail_compilation/named_arguments_ifti_error.d [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/named_arguments_ifti_error.d(17): Error: template `f` is not callable using argument types `!()(int, int)`
```
Comment (RazvanN7) [inline compiler/test/fail_compilation/named_arguments_ifti_error.d:4]: This error message does not seem to accurately explain what the problem is. I would expect something along the lines of: "Missing argument for parameter 'y'".

File: compiler/test/fail_compilation/named_arguments_ifti_error.d [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+fail_compilation/named_arguments_ifti_error.d(19): Error: template `f` is not callable using argument types `!()(int)`
+fail_compilation/named_arguments_ifti_error.d(13):        Candidate is: `f(T, U)(T x, U y)`
+---
+*/
+
+void f(T, U)(T x, U y) {}
+
+void main()
+{
+	f(x: 3, x: 3);
```
Comment (RazvanN7) [inline compiler/test/fail_compilation/named_arguments_ifti_error.d:17]: Is this a typo? Did you mean 'y' for the second 'x' ? Either way, the error for assigning a value to 'x' twice is misleading.

File: compiler/src/dmd/dtemplate.d (hunk 2 above)
Comment (dkorpel) [reply compiler/src/dmd/dtemplate.d:5757]: It's intended, I'm updating outdated or confusing comments as I come across them. (Though I won't mix unrelated code changes of course)

File: compiler/test/fail_compilation/named_arguments_ifti_error.d (hunk 3 above)
Comment (dkorpel) [reply compiler/test/fail_compilation/named_arguments_ifti_error.d:4]: The problem is that the current template instantiation code gags / discards most error messages, requiring a refactor to make more informative errors. I don't want to add that to this PR, but want to look into improving this next (which is long overdue; template errors are a big source of confusion especially with `ref` and lambda parameters).

**Review by RazvanN7 — APPROVED** [review]

**Comment by ibuclaw** [comment]
Is this worth retargeting for stable/the current 2.108 beta?

**Comment by RazvanN7** [comment]
cc @dkorpel

**Comment by dkorpel** [comment]
Yes, this was meant for 2.108, so targeting stable now.

**Comment by thewilsonator** [comment]
Is this supposed to contain 5 commits, 4 of which look unrelated?

**Comment by dkorpel** [comment]
No, I don't know why they are there. Git says it's up to date with stable.
